From: Jens Axboe <axboe@fb.com>
Date: Thu, 19 Jan 2017 07:58:59 -0700
Subject: blk-mq: stop hardware queue in blk_mq_delay_queue()
References: bsc#1020989,FATE#322379
Git-commit: 7e79dadce222e06e0c30a77280f3426014bee185
Patch-Mainline: v4.11-rc1

The run handler we register for the delayed work requires that the
queue be stopped, yet we leave that up to the caller. Let's move
it into blk_mq_delay_queue() itself, so that the API is sane.

This fixes a stall with SCSI, where it calls blk_mq_delay_queue()
without having stopped the queue. Hence the queue is never run.

Reported-by: Hannes Reinecke <hare@suse.com>
Fixes: 70f4db639c5b ("blk-mq: add blk_mq_delay_queue")
Signed-off-by: Jens Axboe <axboe@fb.com>
Signed-off-by: Hannes Reinecke <hare@suse.de>
---
 block/blk-mq.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/block/blk-mq.c b/block/blk-mq.c
index 90b42af..70143ff 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -1117,6 +1117,7 @@ void blk_mq_delay_queue(struct blk_mq_hw_ctx *hctx, unsigned long msecs)
 	if (unlikely(!blk_mq_hw_queue_mapped(hctx)))
 		return;
 
+	blk_mq_stop_hw_queue(hctx);
 	kblockd_schedule_delayed_work_on(blk_mq_hctx_next_cpu(hctx),
 			&hctx->delay_work, msecs_to_jiffies(msecs));
 }
-- 
1.8.5.6

